src/parse.y

```diff
@@ -53,6 +55,30 @@
 	enum node_flavor node_flavor;
 }
 
+%{
+	// local variables and functions
+	static const char *parsing_filename;
+	static struct policy_node *cur;
+	static enum node_flavor expected_node_flavor;
+	static void yyerror(const YYLTYPE *locp, yyscan_t yyscanner, char const *msg);
+
+	// lexer
+	extern void yyrestart(FILE *input_file , yyscan_t yyscanner);
+	extern int yylex(YYSTYPE *yylval_param, YYLTYPE *yylloc_param, yyscan_t yyscanner);
+	extern int yylex_init(yyscan_t* scanner);
+	extern int yylex_destroy(yyscan_t scanner);
+	extern char *current_lines[LINES_TO_CACHE];
+	extern unsigned line_cache_index;
+	extern void reset_current_lines(void);
+%}
+
+%code provides {
+	#define LINES_TO_CACHE 5
```

Comment:
No, not for comments but for multiline statements, e.g.:
```
allow
  source
  target
  :class 
  { perm1 perm2 };
```
See `./tests/functional/policies/parse_errors/test3.output` for an example.
```
test3_tmp.if:         2: (F): Error: Unexpected te-file parsed (F-001)
    2 |     policy_module(          test2,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    3 |
      | ^
    4 |
      | ^
    5 |             0.0.1
      |             ^~~~~
    6 |   )
      |   ^
Error: Failed to parse files
```

In case the statement exceeds 5 lines the output gets truncated, see `./tests/functional/policies/parse_errors/test7.output`:
```
test7.if:             1: (F): Error: Unexpected te-file parsed (F-001)
    6 |  ...  [truncated]
    7 |   ,
      |   ^
    8 |
      | ^
    9 |   0.0.1
      |   ^~~~~
   10 |
      | ^
   11 |   )
      |   ^
Error: Failed to parse files
```

`5` is just a random number I came up with.

Reply:
Okay, that sounds good.  And 5 seems like a reasonable default to me.  Can you add a comment above the define indicating that it controls the number of lines to output before truncation on a parse error?

Reply:
I'm having trouble getting this to work.  I wrote policy that says:

```
allow
foo_t;
```

I was expecting multiline output, but got:

```
[dburgener@localhost refpolicy]$ ../selint/src/selint -rsS -e S-010 -e S-009 -E policy
kernel.te:           25: (F): syntax error, unexpected SEMICOLON (F-001)
   25 | foo_t     ;
      |           ^
Error: Failed to parse files
```

Reply:
Same policy, I now get:

```
[dburgener@localhost refpolicy]$ ../selint/src/selint -rsS policy
kernel.te:           25: (F): syntax error, unexpected SEMICOLON (F-001)
   25 | foo_t;
      |      ^
kernel.te:           24: (F): Error: Invalid statement (F-001)
   24 | allow
      | ^~~~~
   25 | foo_t;
      | ~~~~~~
Error: Failed to parse files
```

I don't even know for sure what sort of error message I would want/expect in this situation.  What's the intended behavior for something like this?

Reply:
@cgzones ping on this.  This error seems a little weird to me.  Curious on your thoughts.

Reply:
With the `allow` keyword and the following string the parser entered the rule `bare_line` and is expecting now either the sub-rule `rule` (with the syntax `av_type string_list string_list COLON string_list string_list SEMICOLON`) or `role_allow` (with the syntax `av_type string_list string_list SEMICOLON`).
Since the actual following `;` is not parsable as a `string_list`, we get an error.

In your first error output there was no `bare_line` error handler:
```
error {
		const struct location loc = { @1.first_line, @1.first_column, @1.last_line, @1.last_column };
		yyerror(&loc, NULL, "Error: Invalid statement");
		YYABORT;
		}
```
so the syntax error propagated all the way to the top, and there arrived we have not information what the exact issue was and how many previous lines are related to this syntax error.

With that error handler, in your second error output, the second part is added
```
kernel.te:           24: (F): Error: Invalid statement (F-001)
   24 | allow
      | ^~~~~
   25 | foo_t;
      | ~~~~~~
```
cause we know where the statement started and can print the corresponding lines.

I am not set on what final error output should be display either.

Reply:
Okay.  That all seems reasonable.

I think that if neither of us are set on how we would *prefer* this sort of error message to look, this is probably fine for now.  Maybe when we have a real world example of a multi-line parse error, it will become clearer what the error output should be.  In the meantime, it seems very reasonable for single line errors, and this output is at the worst case unhelpful, rather than actually being wrong or anything.

In my opinion, I think we should go ahead and merge this PR, and any additional cleanup for cases like this can get added in later as needed.  Any objections on your end, or is this ready to go from your perspective?